com.unity.multiplayer.mlapi/Runtime/Timing/NetworkTimeSystem.cs

```diff
@@ -0,0 +1,111 @@
+        /// </summary>
+        public double HardResetThreshold { get; set; }
+
+        /// <summary>
+        /// Gets or sets the ratio at which the NetworkTimeSystem speeds up or slows down time.
+        /// </summary>
+        public double AdjustmentRatio { get; set; }
+
+        public double LocalTime => m_Time + m_CurrentLocalTimeOffset;
+
+        public double ServerTime => m_Time + m_CurrentServerTimeOffset;
+
+        //TODO This is used as a workaround to pass this back into the sync function will be removed once we get correct value.
+        internal double TimeSystemInternalTime => m_Time;
+
+        public NetworkTimeSystem(double localBuffer, double serverBuffer, double hardResetThreshold, double adjustmentRatio = 0.01d)
+        {
+            LocalBuffer = localBuffer;
+            ServerBuffer = serverBuffer;
+            HardResetThreshold = hardResetThreshold;
+            AdjustmentRatio = adjustmentRatio;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTimeSystem"/> class for a server instance.
+        /// The server will not apply any buffer values which ensures that local time equals server time.
+        /// </summary>
+        /// <returns>The instance.</returns>
+        public static NetworkTimeSystem ServerTimeSystem()
+        {
+            return new NetworkTimeSystem(0, 0, double.MaxValue);
+        }
+
+        /// <summary>
+        /// Advances the time system by a certain amount of time. Should be called once per frame with Time.deltaTime or similar.
+        /// </summary>
+        /// <param name="deltaTime">The amount of time to advance. The delta time which passed since Advance was last called.</param>
+        /// <returns></returns>
+        public bool Advance(double deltaTime)
+        {
+            m_Time += deltaTime;
+
+            if (Math.Abs(m_DesiredLocalTimeOffset - m_CurrentLocalTimeOffset) > HardResetThreshold || Math.Abs(m_DesiredServerTimeOffset - m_CurrentServerTimeOffset) > HardResetThreshold)
+            {
+                m_Time += m_DesiredServerTimeOffset;
+
+                m_DesiredLocalTimeOffset -= m_DesiredServerTimeOffset;
+                m_CurrentLocalTimeOffset = m_DesiredLocalTimeOffset;
+
+                m_DesiredServerTimeOffset = 0;
+                m_CurrentServerTimeOffset = 0;
+
+                return true;
+            }
+
+            m_CurrentLocalTimeOffset += deltaTime * (m_DesiredLocalTimeOffset > m_CurrentLocalTimeOffset ? AdjustmentRatio : -AdjustmentRatio);
+            m_CurrentServerTimeOffset += deltaTime * (m_DesiredServerTimeOffset > m_CurrentServerTimeOffset ? AdjustmentRatio : -AdjustmentRatio);
+
+            return false;
+        }
+
+        public void Initialize(double serverTime, double rtt)
+        {
+            Sync(serverTime, rtt);
+            Advance(0);
+        }
+
+        public void Sync(double serverTime, double rtt)
```

Comment:
nit: Calculate vs Sync? sync sounds like we'd do a network operation or reach out to somewhere external to "sync with" — on the other hand, "calculate" would communicate that we'll do a calculation locally.
```suggestion
        public void Calculate(double serverTime, double rtt)
```

Reply:
So what this does (and I just realized this lacks XML docs, will fix) is it takes up to date network statistic and synchronizes them to the time system. It does not result in a calculation which changes any of the time values. The advance function is what does that. So that's why I think the name `Sync` is more fitting.